protocols/kad/src/behaviour/test.rs

```diff
@@ -120,27 +123,26 @@ fn bootstrap() {
         let expected_known = swarm_ids.iter().skip(1).cloned().collect::<HashSet<_>>();
 
         // Run test
-        current_thread::run(
-            future::poll_fn(move || {
-                for (i, swarm) in swarms.iter_mut().enumerate() {
-                    loop {
-                        match swarm.poll().unwrap() {
-                            Async::Ready(Some(KademliaEvent::BootstrapResult(Ok(ok)))) => {
-                                assert_eq!(i, 0);
-                                assert_eq!(ok.peer, swarm_ids[0]);
-                                let known = swarm.kbuckets.iter()
-                                    .map(|e| e.node.key.preimage().clone())
-                                    .collect::<HashSet<_>>();
-                                assert_eq!(expected_known, known);
-                                return Ok(Async::Ready(()));
-                            }
-                            Async::Ready(_) => (),
-                            Async::NotReady => break,
-                        }
-                    }
+        block_on(async move {
```

Comment:
I find the test easier to understand, when we first poll all swarms expect the first one, giving them a chance to make progress, and then check whether the first swarm has bootstrapped successfully.

What do you think @romanb? I would apply the same pattern to the similar tests below in case people are in favor of this style.

Reply:
I'm generally OK with it, though I don't find it simpler personally, but it seems to be semantically equivalent (i.e. each swarm is polled until pending or the expected result appears). However, I have two small comments:

  1. Is the `'outer` loop label necessary (i.e. why not `return` as before)?
  2. If I see it correctly, then while all swarms are "not ready" / "pending" (e.g. waiting for I/O), this variant seems to enter a "busy loop", not yielding control of the thread to do other useful things meanwhile. While this probably doesn't matter much especially since the tests use the `MemoryTransport`, I still find it cleaner to yield the thread when all swarms are pending.

Reply:
Another small difference to the previous code seems to be that this version does not fail the test if any other swarm besides the first produces a `BootstrapResult` (which would naturally be incorrect and unexpected). Again, this may not be important, but it is a small difference.